### PR TITLE
Add a migration guide for V3

### PR DESCRIPTION
**Pull request by robinmetral**
## Purpose

Add a migration guide for migrating from Circuit UI v2.x to Circuit UI v3.

👉 [**View the rendered MIGRATION.md**](https://github.com/sumup-oss/circuit-ui/blob/383377fc0e09a560dfc9682642f8113b384cc4df/MIGRATION.md) 👈 

## Approach and changes

☝️ 

## Definition of done

* [ ] Development completed
* [ ] Reviewers assigned
* [ ] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

### CODE REVIEW

File: MIGRATION.md [hunk 1]
```diff
@@ -35,6 +45,88 @@ You can only run one codemod at a time and we encourage you to apply the transfo
...
+
+Use 🤖 _body-variant-highlight_ to migrate `bold` to `variant="highlight"`. Note that this will transform your `<p>` elements into inline `<strong>` elements: you might also need to pass `as="p"` if you need a block-level element.
+
+> Bonus: the new `Body` component also supports the `success`, `error` and `subtle` semantic variants. Have a look at the [story](https://circuit.sumup.com/?path=/story/typography-body--variants) to get started.
+
+#### New sizes
+
+(...)
+
+- Components: the `size` prop has been changed to accept the new size numbers for **Headline**, **SubHeadline**, **Body**. For **Headline** component **_exa_** and **_peta_** has been changed to **_one_** with new values, **_tera_** has been changed to **_two_**, **_giga_** to **_three_**, **_mega_** and **_kilo_** to **_four_**. For **Body** component **_giga_** size has been removed, and **_mega_** and **_kilo_** sizes have been changed to **_one_** and **_two_** respectively. The **SubHeadline** component now uses only one size value (🤖 _typography-sizes_).
```
Comment (connor-baer) [inline MIGRATION.md:102]: I wonder if this would be more digestible as a table.

Comment (robinmetral) [reply MIGRATION.md:102]: Yes, it will be one :slightly_smiling_face: Sorry this is still very much a WIP :construction:

File: MIGRATION.md [hunk 2]
```diff
@@ -53,13 +60,59 @@ Tip: Provide the `--transform`/`-t` argument at the end of the command, so that
...
 ## From v2.x to v3
 
-(intro)
+Circuit v3 is a large major release, including long-awaited changes from the full year that passed since v2. This guide will help you upgrade your application. Don't hesitate to contact the [maintainers](https://github.com/sumup-oss/circuit-ui#maintainers) if you have any further questions.
+
+### Accessibility
+
+Any accessible label props that were previously optional are now required and enforced in all components.
+
+To help identify places where accessible labels are missing (even if an app is built with TypeScript, some configurations will not break the build if required props are missing in JavaScript files), components will throw runtime errors in development at missing labels. Production and testing builds are not affected.
```
Comment (connor-baer) [inline MIGRATION.md:69]: The interjection about TypeScript disrupts the flow of the sentence and makes it more difficult to understand. I'd remove the interjection entirely or reword it to separate the what (throwing errors) and why (no type checks in JavaScript and imperfect type checks in TypeScript).

File: MIGRATION.md [hunk 3]
```diff
@@ -53,13 +60,59 @@ Tip: Provide the `--transform`/`-t` argument at the end of the command, so that
...
+
+Now, if you want to turn off the accessibility errors temporarily, run the development app with the environment variable set to `true`:
+
+```sh
+UNSAFE_DISABLE_ACCESSIBILITY_ERRORS=true yarn dev # or yarn start
+```
+
+Keep in mind that this escape hatch is not meant as a way to permanently avoid the errors, but as a temporary workaround while the missing labels are being written, localized and added to the relevant components.
+
+> Note: for the `Input` and `Select` components, use the built-in `label` prop instead of using the `Label` component separately.
```
Comment (connor-baer) [inline MIGRATION.md:83]: This seems out of place. This change was already done in v2. How about this:

```suggestion
> Reminder: For the `Input` and `Select` components, use the built-in `label` prop instead of using the `Label` component separately.
```

File: MIGRATION.md [hunk 4]
```diff
@@ -53,13 +60,59 @@ Tip: Provide the `--transform`/`-t` argument at the end of the command, so that
...
+
+Other accessibility improvements include:
+
+- Accessible modals and popovers, see [the dedicated section](#consolidated-modals-and-popovers)
+- Semantic heading elements, see [the typography section](#heading-as-props)
+
+### New JSX transform
+
+Circuit v3 improves compatibility with the [new JSX transfors](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) (introduced inReact v17).
```
Comment (connor-baer) [inline MIGRATION.md:92]:
```suggestion
Circuit v3 improves compatibility with the [new JSX transforms](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) (introduced in React v17).
```

File: MIGRATION.md [hunk 5]
```diff
@@ -53,13 +60,59 @@ Tip: Provide the `--transform`/`-t` argument at the end of the command, so that
...
+        }
+      }
+    ]
+  ],
+  "plugins": [["babel-plugin-emotion", { "cssPropOptimization": true }]]
+}
+```
 
-(...)
+See the [Emotion 10 docs](https://5faaafd0bd0f3f0008469537--emotion.netlify.app/docs/css-prop#babel-preset) for more details on the required config.
```
Comment (connor-baer) [inline MIGRATION.md:113]: Is this required to use Circuit UI or just to use Emotion with the `css` prop in the application?

I'd assume that it's not required for Circuit UI since it's pre-transpiled with these settings.

File: MIGRATION.md [hunk 6]
```diff
@@ -130,7 +187,7 @@ The deprecated `text[Kilo|Mega|Giga]` style mixins were replaced by a single `ty
 
 #### Heading `as` props
 
-The `as` prop is now required in both the **Headline** and the **SubHeadline** components, and restricted to HTML heading elements. This makes heading structure explicit, and ensures that heading components render semantic heading elements.
+The `as` prop is now required in both the **Headline** and the **SubHeadline** components. Intentionally setting the heading level ensures a consistent and accessible page structure. The `as` prop values was also restricted to HTML heading elements to ensures that heading components render semantic heading elements.
```
Comment (connor-baer) [inline MIGRATION.md:190]:
```suggestion
The `as` prop is now required in both the **Headline** and the **SubHeadline** components. Intentionally setting the heading level ensures a consistent and accessible page structure. The `as` prop values were also restricted to HTML heading elements to ensure that heading components render semantic heading elements.
```

File: MIGRATION.md [hunk 7]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
 
 > Generally, avoid using typography size tokens. Instead, use typography components directly.
 
-### Modal improvements
+### Consolidated modals and popovers
```
Comment (connor-baer) [inline MIGRATION.md:207]:
```suggestion
### New modal and popover APIs
```

File: MIGRATION.md [hunk 8]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
+
+The `Popover` component was rebuilt in Circuit v3. It now uses [Popper v2](https://popper.js.org/) under the hood and comes with a refreshed component API.
+
+More importantly, the Circuit v3 Popover is opinionated when it comes to its usage pattern. It no longer accepts custom children, unlike v2, but rather a list of actions (links and/or buttons).
+
+To migrate, separate the popovers in your application into two types.
+
+Start by migrating any popover resembling a dropdown menu to the new Circuit v3 Popover. Refer to [the Popover story](https://circuit.sumup.com/?path=/story/components-popover--base) for a usage example.
+
+Most of the remaining popovers with custom children should gradually be migrated to use different UI patterns (for example modals). In the meantime, we recommend creating a local copy of the Circuit v3 Popover in your application that accepts custom children instead of an array of actions. You'll find [an example](https://github.com/sumup/ze-dashboard/blob/master/src/components/Popover/Popover.tsx) in the SumUp merchant dashboard (private repository).
```
Comment (connor-baer) [inline MIGRATION.md:229]: Again, not happy about the link to internal code, though I can't think of an alternative. I like the `(private repository)` label.

File: MIGRATION.md [hunk 9]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
+| Select                        |        40px        |        48px        |
+| Tabs                          |        80px        |        48px        |
+| Tag                           |        34px        |        32px        |
+
+We recommend verifying these changes visually at the end of the migration.
 
-- enforced labels
-- throwing errors + escape hatch
-- as prop in headlines (also mentioned in typography)
+In addition to its increased height, the `Button`'s default size was renamed from `mega` to `giga` to align it with the new size values (see the table above). (🤖 button-default-size)
```
Comment (connor-baer) [inline MIGRATION.md:253]:
```suggestion
In addition to its increased height, the `Button`'s default size was renamed from `mega` to `giga` to align it with the new size values (see the table above). (🤖 _button-default-size_)
```

File: MIGRATION.md [hunk 10]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
 
 ### Other changes
 
 - The **NotificationBanner** component has been renamed to **NotificationCard**. (🤖 _component-names-v3_)
-- The **TableRow**, **TableHeader** and **TableCell** components are not exported anymore. Use the **Table** component instead.
-- The **Button**, **IconButton**, (...) and **Hamburger** components' dimensions have been harmonized.
-- The **SelectorGroup**'s `label` is now visible by default, pass `hideLabel` to hide it visually.
-- (...)
+- Label prop names across components were harmonized to follow the _actionLabel_ pattern. (🤖 label-prop-names)
```
Comment (connor-baer) [inline MIGRATION.md:258]:
```suggestion
- Label prop names across components were harmonized to follow the _<description>Label_ pattern. (🤖 _label-prop-names_)
```

File: MIGRATION.md [hunk 11]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
+  - **CardHeader**: `labelCloseButton` 👉 `closeButtonLabel`
+  - **Hamburger**: `labelActive` 👉 `activeLabel`, `labelInActive` 👉 `inactiveLabel`
+  - **Tag**: `labelRemoveButton` 👉 `removeButtonLabel`
+  - **Toggle**: `labelChecked` 👉 `checkedLabel`, `labelUnchecked` 👉 `uncheckedLabel`
+- The **TableRow**, **TableHeader** and **TableCell** components are no longer exported. Use the **Table** component instead.
+- The **Table**'s custom `onSortBy` method signature has been changed. The `nextDirection` argument moved to the third position (`(index, nextDirection, rows)` 👉 `(index, rows, nextDirection)`) and is now optional (i.e. it can be `undefined` instead of `null` in the previous implementation).
+- The **SelectorGroup**'s `label` is now visible by default, pass `hideLabel` to hide it visually. Its children are now also rendered horizontally by default.
```
Comment (connor-baer) [inline MIGRATION.md:265]:
```suggestion
- The **SelectorGroup**'s `label` is now visible by default, pass `hideLabel` to hide it visually. Its children are now rendered horizontally by default.
```

File: MIGRATION.md [hunk 12]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
+- Default `data-testids` are no longer built into the **Table** and **CardHeader** components. We recommend [querying by role](https://testing-library.com/docs/queries/about/#priority) in tests instead, for them to resemble how users interact with our applications.
```
Comment (connor-baer) [inline MIGRATION.md:266]:
```suggestion
- Default `data-testids` are no longer built into the **Table** and **CardHeader** components. We recommend [querying by role](https://testing-library.com/docs/queries/about/#priority) in tests instead to imitate how users interact with our applications.
```

File: MIGRATION.md [hunk 13]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
+
+### Cleaning up
+
+Finally, Circuit v3 removes previously deprecated and/or unused features and props. These breaking changes may not affect your application if you've already addressed the deprecation warnings in Circuit v2 minors, but we still recommend going through the list of changes below.
+
+- The deprecated **Spacing** component has been removed. Use the **spacing** style mixin instead.
+- The **ProgressBar**'s deprecated `children` prop has been removed. Use the label prop instead.
```
Comment (connor-baer) [inline MIGRATION.md:273]:
```suggestion
- The **ProgressBar**'s deprecated `children` prop has been removed. Use the `label` prop instead.
```

File: MIGRATION.md [hunk 14]
```diff
@@ -147,23 +204,80 @@ There is no codemod for these changes, migrate manually through search and repla
...
+- The **Card**'s deprecated `shadow` prop has been removed. Shadows have been replaced by a single outline in an earlier minor version.
+- The deprecated `styleHelpers` aggregate is no longer exported. Import each style mixin directly instead.
+- The `themePropType` is no longer exported from `@sumup/circuit-ui`. Import it from `@sumup/design-tokens` instead.
+- The deprecated **withComponents** HOC has been removed. Use the **useComponents** hook instead.
```
Comment (connor-baer) [inline MIGRATION.md:277]:
```suggestion
- The deprecated `withComponents` HOC has been removed. Use the `useComponents` hook instead.
```

**Edit**: Mhm, I see that you/we've used bold for component names and code for prop names. Not sure what to use here 🤔

**Review by connor-baer — COMMENTED** [review]
Wow, I didn't quite realize how many changes and improvements we've made until I saw them listed here! 🤯 

Thank you for documenting all of them so thoroughly 🙌🏻

**Review by amelako — APPROVED** [review]
Looks great. No additional comments. 🚀

File: MIGRATION.md (hunk 4 above)
Comment (robinmetral) [reply MIGRATION.md:92]: Oops 🙈

File: MIGRATION.md (hunk 5 above)
Comment (robinmetral) [reply MIGRATION.md:113]: Ah, maybe that's only to use Emotion with `css`. Would you recommend simplifying the example config?

File: MIGRATION.md (hunk 6 above)
Comment (robinmetral) [reply MIGRATION.md:190]: Oh my, I think I was getting a bit tired by this point 🙈

File: MIGRATION.md (hunk 8 above)
Comment (robinmetral) [reply MIGRATION.md:229]: Agreed, but should we leave it in this case if we don't have an alternative? I think it could be helpful to people who have access to see how the local component was implemented.

File: MIGRATION.md (hunk 14 above)
Comment (robinmetral) [reply MIGRATION.md:277]: Same 😄 I was also unsure so I "decided" to make any exports **bold** and props `code`. What do you think?

**Comment by robinmetral** [comment]
Thanks for reviewing @connor-baer @amelako! I addressed most comments and commented in the rest ☝️

File: MIGRATION.md (hunk 14 above)
Comment (connor-baer) [reply MIGRATION.md:277]: Works for me.

File: MIGRATION.md (hunk 8 above)
Comment (connor-baer) [reply MIGRATION.md:229]: Yes, let's leave it with the label.

**Review by connor-baer — APPROVED** [review]
